src/lib/testing/utils.ts

```diff
@@ -1,17 +1,42 @@
 import { createDb, type DB } from '$lib/server';
+import { tables } from '$lib/server/db/schema';
 import { PostgreSqlContainer } from '@testcontainers/postgresql';
+import { Mutex } from 'async-mutex';
+import { test, type ExpectStatic } from 'vitest';
 
-export const createTestingDb = async <T>(generator?: (db: DB) => Promise<T>) => {
-    const container = new PostgreSqlContainer();
-    const startedContainer = await container.start();
-    const db = await createDb(startedContainer.getConnectionUri());
+type DbStateGenerator<T> = (db: DB) => Promise<T>;
+type TestFunc<T> = (args: {
+    db: DB;
+    generated?: Awaited<T>;
+    expect: ExpectStatic;
+}) => Promise<void>;
 
-    const generated = generator ? await generator(db) : null;
-
-    return { container: startedContainer, db, generated };
-};
+const container = new PostgreSqlContainer();
+const startedContainer = await container.start();
+const db = await createDb(startedContainer.getConnectionUri());
+const lock = new Mutex();
 
 export const mustGenerate = <T>(generated: T | null) => {
     if (!generated) process.exit(1);
     return generated;
 };
+
+export const testWithDb = <T>(
+    name: string,
+    testFunc: TestFunc<T>,
+    generator?: DbStateGenerator<T>
+) => {
+    test(name, async ({ expect }) => {
+        await lock.runExclusive(async () => {
+            await db.transaction(async (tx) => {
+                for (const table of tables) {
+                    await tx.delete(table);
+                }
+            });
+
+            const generated = generator ? await generator(db) : undefined;
+
+            await testFunc({ db, generated, expect });
+        });
+    });
```

Comment:
Would be nice to have access to the full object that is passed by vitest to the function in `test`.
i.e.,
```ts
test(name, async obj => {
    ...
    await testFunc({ db, generated, ...obj });
})
```

Reply:
The type of obj seems to be weird such that `expect` isn't able to be destructured from it after the fact. Something breaks when I do this.

Reply:
I played with it a bit.  We can get this to work by doing
```ts
const newArgs = args as TestArgs<T>;
newArgs.db = db;
newArgs.generated = generated;

await testFunc(newArgs);
```
where `TestArgs` is
```ts
type TestArgs<T> = { db: DB; generated?: Awaited<T>; } & ExtendedContext<RunnerTestCase> & TestContext;
```

Reply:
My best guess is that `args` is a class with some funkiness for getting the `expect` or maybe it's using Proxies some how

Reply:
Yeah it is kinda bizarre. What you suggested works though

Reply:
Wonderful